src/storage/src/render/sources.rs

```diff
@@ -320,12 +320,43 @@ where
                     let transformed_results =
                         transform_keys_from_key_envelope(upsert_envelope, results);
 
+                    let persist_clients = Arc::clone(&storage_state.persist_clients);
+
+                    let previous_as_of = match resume_upper.as_option() {
+                        None => Some(Timestamp::MAX),
+                        Some(&0) => None,
+                        Some(&t) => Some(t - 1),
+                    };
+                    let (previous_ok_stream, previous_err_stream, previous_token) =
+                        if let Some(previous_as_of) = previous_as_of {
+                            let (ok, err, tok) = persist_source::persist_source(
+                                scope,
+                                persist_clients,
+                                description.storage_metadata.clone(),
+                                Antichain::from_elem(previous_as_of),
+                            );
+                            (ok, err, Some(tok))
+                        } else {
+                            (
+                                std::iter::empty().to_stream(scope),
+                                std::iter::empty().to_stream(scope),
+                                None,
+                            )
+                        };
+                    let (previous_ok, previous_err) = (
+                        previous_ok_stream.as_collection(),
+                        previous_err_stream.as_collection(),
+                    );
+                    error_collections.push(previous_err);
```

Comment:
I may discover this later, but .. does this mean we are unable to recover from upsert errors?